src/KMonad/App/KeyIO/Mac/IOKitSource.hs

```diff
@@ -0,0 +1,72 @@
+{-|
+Module: KMonad.App.KeyIO.Mac.IOKitSource
+Description: Use mac c-api to grab keyboard(s).
+TODO: complete this
+-}
+module KMonad.App.KeyIO.Mac.IOKitSource
+  ( withIOKitSource )
+where
+
+import KMonad.Prelude
+
+import Foreign.C.String
+import Foreign.Ptr
+import Foreign.Marshal hiding (void)
+import Foreign.Storable
+
+import KMonad.App.KeyIO.Common
+import KMonad.Util
+import KMonad.Util.Keyboard.Mac
+
+
+-------------------------------------------------------------------------------
+-- $ffi
+
+-- | Use the mac c-api to grab keyboard(s)
+foreign import ccall "grab_kb"
+  grab_kb :: CString -> OnlyIO Word8
+
+-- | Release the keyboard hook
+foreign import ccall "release_kb"
+  release_kb :: OnlyIO Word8
+
+-- | Pass a pointer to a buffer to wait_key, and when it
+-- returns, the buffer can be read for the next key event.
+foreign import ccall "wait_key"
+  wait_key :: Ptr RawEvent -> OnlyIO Word8
+
+
+-------------------------------------------------------------------------------
+
+withIOKitSource :: LUIO m e => IOKitCfg -> Ctx r m GetKey
+withIOKitSource cfg = mkCtx $ \f -> do
+
+  let prodStr = cfg^.productStr
+
+  let init = do
+        case (prodStr) of
+          Nothing -> do
+            logInfo "Opening HID devices"
+            liftIO . void $ grab_kb nullPtr
+          Just s -> do
+            logInfo $ "Opening HID device: " <> s
+            liftIO $ do
+              str <- newCString $ unpack s
+              void $ grab_kb str
+              free str
```

Comment:
Everytime you have this `newCString` -> do stuff -> `free` pattern going on its worth looking at whether `withCString` may be a better fit